File: Python/Worksheet.py
#!/usr/bin/env python
import re
class RegWhy:
    Period="\."
    BackSlash="\\\\"
    Bar ="\|"
    LeftParenthesis="\("
    RightParenthesis= "\)"
    NullString="\0"
    LeftBracket="\["
    RightBracket="\]"
    LeftBrace="\{"
    RightBrace="\}"
    DollarSign="\$"
    Asterik="\*"
    QuestionMark="\?"
    LeftAngle="\<"
    RightAngle="\>"
    Caret="\^"
    PlusSign="+"
    Tab="\t"
    ReturnString="\r"
    NewLine="\n"
    FormFeed="\f"
    OrMarker="|"
    def __init__(self):
        self.CharacterType = self.CharacterType()
        self.Group = self.Group()
        self.Where=self.Where()
        self.Count=self.Count()

    @staticmethod
    def Literal(value):
        finalValue=value
        finalValue=finalValue.replace("\\",RegWhy.BackSlash)
        finalValue=finalValue.replace("." , RegWhy.Period)
        finalValue=finalValue.replace("|",RegWhy.Bar)
        finalValue=finalValue.replace("(",RegWhy.LeftParenthesis)
        finalValue=finalValue.replace(")",RegWhy.RightParenthesis)
        finalValue=finalValue.replace("[",RegWhy.LeftBracket)
        finalValue=finalValue.replace("]",RegWhy.RightBracket)
        finalValue=finalValue.replace("{",RegWhy.LeftBrace)
        finalValue=finalValue.replace("}",RegWhy.RightBrace)
        finalValue=finalValue.replace("$",RegWhy.DollarSign)
        finalValue=finalValue.replace("*",RegWhy.Asterik)
        finalValue=finalValue.replace("?",RegWhy.QuestionMark)
        finalValue=finalValue.replace("<",RegWhy.LeftAngle)
        finalValue=finalValue.replace(">",RegWhy.RightAngle)
        finalValue=finalValue.replace("^",RegWhy.Caret)
        finalValue=finalValue.replace("+",RegWhy.PlusSign)
        return finalValue
    @staticmethod
    def Statement(listOfRegWhy):
            final_string=""
            for str in listOfRegWhy:
                final_string +=str
            return final_string
    
    class CharacterType:
        AnyCharacter="."
        Digit ="\d"
        NonDigit="\D"
        WhiteSpace="\s"
        NonWhiteSpace="\S"
        WordCharacter="\w"
        NonWordCharacter="\W"
        WordBounday="\b"
        NonWordBoundary="\B"
        LowerCaseASCI="[a-z]"
        UpperCaseASCII="[A-Z]"
        AnyASCII='[ -~]'
        EndOfFile="^Z"
        @staticmethod
        def Unicode(fourDigitUniCodeNumber):
            unicodeString ="\\u" + fourDigitUniCodeNumber
            return (unicodeString)

        @staticmethod
        def Hexidecimal(twoDigitNumberCode):
            return ("\\" +"x" + twoDigitNumberCode)

        @staticmethod
        def CharacterRange(listOfCharacters):
            return "[" + listOfCharacters + "]"

        @staticmethod
        def NotInCharacterRange(listOfCharacters):
            return "[^" + listOfCharacters + "]"

        def __init__(self):
            pass
    class Group:
        StartCapturing="("
        StartNonCapturing="(:"
        End=")"
        EndOptional=")?"
        EndZeroOrMore=")*"
        EndOneOfMore=")+"
        SmallestMatch="?"
        LargestMatch=""
        EndOptionSmallestMatch=")??"
        EndZeroOrMoreSmallestMatch=")*?"
        EndOneOrMoreSmallestMatch=")+?"
        EndOfNumberedSmallestMatch="?"
        LastMatch="$&"
        LastParen="$+"
        PrecedingMatch ="%`"
        @staticmethod
        def StartNamedGroup(nameOfGroup):
            return "(<" + nameOfGroup + ">" 
        
        @staticmethod
        def BackReference(whichReference):
            if (whichReference > 9 or whichReference < 1):
                raise Exception("You must have a number between 1 and 9")
            return "$" + whichReference
        
        @staticmethod
        def MatchOnlyIfThisIsNext(whatsNext):
            return "(?=" + whatsNext

        @staticmethod
        def MatchOnlyIfThisIsNotNext(whatsNext):
            return "(?|" + whatsNext

        @staticmethod
        def EndExactNumbe(countOfTimes):
            return "){" + countOfTimes + "}"

        @staticmethod
        def EndExactNumberOfAbove(countOfTimes):
            return "){" + countOfTimes + ",}"

        @staticmethod
        def EndRangeOfTimes(bottomNumber:int,topNumber):
            return "){" + bottomNumber + "," + topNumber +  "}"

        @staticmethod
        def List(listOfItems:list, doCapture:bool):
            finalString="("
            firstitem=True
            if (doCapture):
                finalString=finalString + "?:"
            for item in listOfItems:
                if (firstitem):
                    firstitem=False
                    finalString=finalString + item
                else:
                    finalString ="|" + item
                finalString=finalString + item
            finalString =finalString + ")"
            return finalString
         

        def __init__(self):
            pass

    class Where:
        StartOfString="^"
        EndOfString="$"
        def __init__(self):
            pass
    class Count:
        Optional="?"
        ZeroOrMore="*"
        OneOrMore="+"

        @staticmethod
        def ExactNumber(numberOfDigits):
            return "{" + str(numberOfDigits) + "}"
        
        @staticmethod
        def ExactNumberOrMore(numberOfDigits):
            return "{" + str(numberOfDigits) + ",}"

        @staticmethod
        def RangeOfTimes(bottomNumber,topNumber):
            return "{" + str(bottomNumber) + "," + str(topNumber) + "}"
        
        def __init__(self):
            pass
    class Do:
        def __init__(self):
            pass

        @staticmethod
        def Detect(stringToSearch, valueToFind, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToFind)
            match =regEx.search(stringToSearch)
            if match:
                return True
            else:
                return False

        @staticmethod
        def SetRegEx(caseInsensitive, multilineMatching, valueToFind):
            flags = 0
            if multilineMatching:
                flags = re.MULTILINE
            if caseInsensitive:
                flags |= re.IGNORECASE
            regEx = re.compile(valueToFind, flags)
            return regEx

        @staticmethod
        def ExtractFirst(stringToSearch, valueToFind, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToFind)
            found = regEx.search(stringToSearch)

            return found.group()
        
        @staticmethod
        def ExtractAll(stringToSearch, valueToFind, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToFind)
            found = regEx.findall(stringToSearch)

            return found

        @staticmethod
        def ExtractGroup(stringToSearch, valueToFind,groupNumber:int, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToFind)
            found = regEx.search(stringToSearch)

            return found.group(groupNumber)

        @staticmethod
        def LocateFirst(stringToSearch, valueToFind, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToFind)
            found = regEx.search(stringToSearch)
            return found.start()


        @staticmethod
        def LocateAll(stringToSearch, valueToFind, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToFind)
            found = [m.start(0) for m in regEx.finditer(stringToSearch)]

            return found

        @staticmethod
        def ReplaceFirst(stringToSearch, valueToFind, valueToReplace, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToFind)
            return regEx.sub(valueToReplace, stringToSearch,1)


        @staticmethod
        def ReplaceAll(stringToSearch, valueToFind,valueToReplace, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToFind)
            return regEx.sub(valueToReplace,stringToSearch)

        @staticmethod
        def SplitList(stringToSearch, valueToSplitWith, caseInsensitive=False, multilineMatching=False):
            regEx = RegWhy.Do.SetRegEx(caseInsensitive, multilineMatching, valueToSplitWith)
            return re.split(regEx,stringToSearch)






if __name__ == '__main__':

    print(RegWhy.Statement([
        RegWhy.Group.StartCapturing,
        RegWhy.Literal("\\?+."),
        RegWhy.OrMarker,
        RegWhy.Literal("or"),
        RegWhy.Group.End,
        RegWhy.CharacterType.AnyCharacter,
        RegWhy.CharacterType.Unicode("1234"),
        RegWhy.Count.ExactNumber(3)
    ]))
    print(RegWhy.Do.Detect("test","T",True))
    print(RegWhy.Do.ExtractFirst("test","t"))
    print(RegWhy.Do.ExtractAll("test", "t"))
    print(RegWhy.Do.LocateAll("test", "t"))
    print(RegWhy.Do.ReplaceFirst("test", "t", "q"))
    print(RegWhy.Do.ReplaceAll("test", "t","q"))
    print(RegWhy.Do.SplitList("tester", "t"))

File: TypeScript/Worksheet.ts
abstract class RegWhy {
    constructor() {
        
    }
    public static Period="\.";
    public static BackSlash="\\";
    public static Bar ="\|";
    public static LeftParenthesis="\(";
    public static RightParenthesis="\)";
    public static NullString="\0";
    public static LeftBracket="\[";
    public static RightBracket="\]";
    public static LeftBrace="\{";
    public static RightBrace="\}";
    public static DollarSign="\$";
    public static Asterik="\*";
    public static QuestionMark="\?";
    public static LeftAngle="\<";
    public static RightAngle="\>";
	public static PlusSign="\+"
    public static Caret="\^";
    public static Tab="\t";
    public static ReturnString="\r";
    public static NewLine="\n";
    public static FormFeed="\f";
    public static OrMarker="|";
    public static AlphaNumericCharacterRangePlus(otherCharacters){
        return "[A-Za-z0-9" + otherCharacters + "]";
    }
    public static  NumericCharacterRangePlus( otherCharacters)
            {
                return "[0-9" + otherCharacters + "]";
            }

    public static  Unicode( fourDigitalNumber)
    {
        return "\\\\" + fourDigitalNumber;
    }

    public static  Hexidecimal( twoDigitalCode)
    {
        return "\\\\" + "x" + twoDigitalCode;
    }

    public static  CharacterRange( listOfCharacters)
    {
        return "[" + listOfCharacters + "]";
    }

    public static  NotInCharacterRange( listOfCharacters)
    {
        return "[^" + listOfCharacters + "]";
    }

    public static Statement(listOfStatements: Array<string>) {
        let finalStatement="";
        for ( let statement of listOfStatements){
             
            finalStatement=finalStatement + statement;
        }
        return finalStatement;

    }
    public static Literal(value: string) {
        value = value.replace("\\", RegWhy.BackSlash);
        value = value.replace(".", RegWhy.Period);
        
        value = value.replace("|", RegWhy.Bar);
        value = value.replace("(", RegWhy.LeftParenthesis);
        value = value.replace(")", RegWhy.RightParenthesis);
        value = value.replace("[", RegWhy.LeftBracket);
        value = value.replace("]", RegWhy.RightBracket);
        value = value.replace("{", RegWhy.LeftBrace);
        value = value.replace("}", RegWhy.RightBrace);
        value = value.replace("$", RegWhy.DollarSign);
        value = value.replace("*", RegWhy.Asterik);
        value = value.replace("?", RegWhy.QuestionMark);
        value = value.replace("<", RegWhy.LeftAngle);
        value = value.replace(">", RegWhy.RightAngle);
        value = value.replace("+", RegWhy.PlusSign);
        value = value.replace("^", RegWhy.Caret);
        return value;
    }


    public static CharacterType() {
        return new CharacterType();
    }
    public static Do() {
        return new Do();
    }
    public static Group() {
        return new Group();
    }
    public static Where() {
        return new Where();
    }
    public static Count(){
        return new Count();
    }
     
    
}
 

class CharacterType{
    
    constructor() {
        
    }
    AnyCharacter="."
    Digit ="\d";
    NonDigit="\D";
    Whitespace="\s";
    NonWhitespace="\S";
    WordCharacter="\w";
    NonWordCharacter="\W";
    WordBounday="\b";
    NonWordBoundary="\B";
    LowerCaseASCI="[a-z]";
    UpperCaseASCII="[A-Z]";
    AnyASCII='[ -~]';
    EndOfFile="^Z";
     
    
    public Unicode(fourDigitUniCodeNumber:string):string{
        return ("\\" +"u" + fourDigitUniCodeNumber) ; 
    }
    public Hexidecimal(twoDigitalCode:string):string{
        return ("\\" +"x" + twoDigitalCode) ; 
    }
    public CharacterRange(listOfCharacters){
        return "[" + listOfCharacters + "]";
    }
    public NotInCharacterRange(listOfCharacters){
        return "[^" + listOfCharacters + "]";
    }
    public AlphaNumeric="[A-Za-z0-9]"
    public Printable ="[A-Za-z0-9 .\\`~!@#$%^&*()_-+={[}]|;:'\"<,>?/";
    public Punctuation="[.\\`~!@#$%^&*()_-+={[}]|;:'\"<,>?/]"
    public Space="[\r\n\t \f]"
    public BlankSpace="[ \t]"
    public AlphaNumericCharacterRangePlus(otherValues){
        return "[A-Za-z0-9" + otherValues + "]"
    }
     
}
class Group{

    constructor() {
        
    }
     
    StartCapturing="("
    StartNonCapturing="(?:"
    
    public StartNamed(nameOfGroup){
        return "(<" + nameOfGroup + ">";
    }
    public StartOptionalCapturingList(listOfWords:Array<string>):string{
        let finalStatement="(";
        for ( let statement of listOfWords){
             if (finalStatement.length > 1){
                finalStatement=finalStatement + "|"
             }
            finalStatement=finalStatement + statement;
        }
        return finalStatement +")";
    }
    public StartOptionalNonCapturingList(listOfWords:Array<string>):string{
        let finalStatement="(?:";
        for ( let statement of listOfWords){
             if (finalStatement.length > 2){
                finalStatement=finalStatement + "|"
             }
            finalStatement=finalStatement + statement;
        }
        return finalStatement +")";
    }
    End=")";
    EndOptional=")?";
    EndZeroOrMore=")*";
    EndOneOfMore=")+";
    SmallestMatch="?";
    LargestMatch="";
    EndOptionSmallestMatch=")??";
    EndZeroOrMoreSmallestMatch=")*?";
    EndOneOrMoreSmallestMatch=")+?";
    EndOfNumberedSmallestMatch="?";

    public BackReference(whichReference:number):string{
        if (whichReference > 9 || whichReference < 1){
            throw new Error("Backreferences must be between 1 and 9")
        }
        return "$" + whichReference;
    }
    public LastMatch="$&";
    public LastParen="$+";
    public PrecedingMatch ="%`";


    public MatchOnlyIfThisIsNext(whatsNext){
        return "(?=" + whatsNext
    }
    public MatchOnlyIfThisIsNotNext(whatsNext){
        return "(?|" + whatsNext
    }
    public EndExactNumber(countOfTimes){
        return "){" + countOfTimes + "}";
    }
    public EndExactNumberOrAbove(countOfTimes){
        return "){" + countOfTimes + ",}";
    }
    public EndRange(bottomNumber, topNumber){
        return "){" + bottomNumber + "," + topNumber +  "}";
    }

    
     
}




class Where{

    constructor() {
        
    }
    StartOfString="^";
    EndOfString="$";
     
}
class Do{
    
    constructor() {
        
    }
    public static SetRegExOptions(global:boolean,caseInsensitive:boolean,multilineMatching:boolean):string{
        var returnString="";
        if (global){
            returnString="g";
        }
        if (caseInsensitive){
            returnString =returnString + 'i';
        }
        if (multilineMatching){
            returnString=returnString + "m";
        }
        return returnString;

    }
    public Detect(stringToSearch:string, valueToFind:string, caseInsensitive=false, multilineMatching=false){
        var pattern =new RegExp(valueToFind,Do.SetRegExOptions(false,caseInsensitive,multilineMatching))
        return (pattern.test(stringToSearch))
    }
    public ExtractFirst(stringToSearch:string, valueToFind:string, caseInsensitive=false, multilineMatching=false){
        var pattern =new RegExp(valueToFind,Do.SetRegExOptions(false,caseInsensitive,multilineMatching))
        var result= (pattern.exec(stringToSearch))
        if (result==null){return null}
        return result[0];
    }
    public ExtractAll(stringToSearch:string, valueToFind:string, caseInsensitive=false, multilineMatching=false){
        var matches = [];
        var match;
         
        var pattern =new RegExp(valueToFind,Do.SetRegExOptions(true,caseInsensitive,multilineMatching))
        while (match = pattern.exec(stringToSearch)) {
         matches.push(match[0]);
          
        }
        return matches;
    }
   
    public ExtractCapturedGroup(stringToSearch:string, valueToFind:string, groupToCapture:number, caseInsensitive=false, multilineMatching=false){
        groupToCapture || (groupToCapture = 1); // default to the first capturing group
        var matches = [];
        var match;
        var pattern =new RegExp(valueToFind,Do.SetRegExOptions(true,caseInsensitive,multilineMatching))
        while (match = pattern.exec(stringToSearch)) {
        matches.push(match[0]);
        }
        return matches[groupToCapture];
         
    }
    public LocateFirst(stringToSearch:string, valueToFind:string, caseInsensitive=false, multilineMatching=false){
        var pattern =new RegExp(valueToFind,Do.SetRegExOptions(false,caseInsensitive,multilineMatching))
        
        var result= (pattern.exec(stringToSearch))
        if (result==null){return -1}
        return result.index;
    }
    public LocateAll(stringToSearch:string, valueToFind:string, caseInsensitive=false, multilineMatching=false){
        var matches = [];
        var match;
        var pattern =new RegExp(valueToFind,Do.SetRegExOptions(true,caseInsensitive,multilineMatching))
         
        while (match = pattern.exec(stringToSearch)) {
            matches.push(match.index);
            }
         
        return matches;
    }
    public ReplaceFirst(stringToSearch:string, valueToFind:string, valueToReplaceWith:string, caseInsensitive=false, multilineMatching=false):string{
        var pattern =new RegExp(valueToFind,Do.SetRegExOptions(false,caseInsensitive,multilineMatching))
        return stringToSearch.replace(pattern,valueToReplaceWith);
    }
    public ReplaceAll(stringToSearch:string, valueToFind:string,valueToReplaceWith:string, caseInsensitive=false, multilineMatching=false){
        var pattern =new RegExp(valueToFind,Do.SetRegExOptions(true,caseInsensitive,multilineMatching))
        return stringToSearch.split(pattern).join(valueToReplaceWith)
    }
    public SplitList(stringToSearch:string, valueToSplitWith:string, caseInsensitive=false, multilineMatching=false){
        var pattern =new RegExp(valueToSplitWith,Do.SetRegExOptions(false,caseInsensitive,multilineMatching))
        return stringToSearch.split(pattern);
    }
   
}
class Count{
    constructor(){

    }
    Optional="?";
    ZeroOrMore="*";
    OneOrMore="+";
    public ExactNumber(numberOfDigits:number){
        return "{" + numberOfDigits + "}";
    }
    public ExactNumberOrMore(numberOfDigits:number){
        return "{" + numberOfDigits + ",}";
    }
    public RangeOfTimes(bottomNumber:number, topNUmber:number){
        return "{" + bottomNumber + "," + topNUmber + "}";
    }
}
// var regWhy = new RegWhy();

//  console.log("testin")
// console.log(regWhy.Do().ExtractAll("Test","t"));
console.log(RegWhy.Statement(["test","test"]))
console.log(RegWhy.Do().ExtractAll("test", 't'));
console.log(RegWhy.Do().ExtractFirst("test", 't'));
console.log(RegWhy.Do().LocateFirst("test", 't'));
console.log(RegWhy.Do().LocateAll("test", 't'));
console.log(RegWhy.Literal("hello world"));
console.log(RegWhy.Do().Detect("test","t"))
console.log(RegWhy.Do().Detect("test","x"))
console.log(RegWhy.Statement([
    RegWhy.Group().StartNonCapturing,
  RegWhy.Literal("INT"),
  RegWhy.OrMarker,
  RegWhy.Literal("EXT"),
  RegWhy.Group().End,
  RegWhy.Period,
  RegWhy.Count().ZeroOrMore,
  RegWhy.CharacterType().Whitespace,
  RegWhy.Count().OneOrMore,
  RegWhy.Group().StartCapturing,
  RegWhy.CharacterType().AlphaNumeric,
  RegWhy.Count().OneOrMore,
  RegWhy.Group().End
]))
 

